L64-python-runtime-proto-parsing.md

```diff
@@ -0,0 +1,99 @@
+    pass
+```
+
+The final function, `protos_and_services` is a simple convenience function allowing the user to import protos and services in a single function call.
+
+The change will be entirely backward compatible. Users manually invoking `protoc` today will not be required to change their build process.
+
+### Import Paths
+
+These functions will behave like normal `import` statements. `sys.path` will be used to search for `.proto` files. The path under which each particular `.proto` file was found will be passed to the protobuf parser as the root of the tree (equivalent to the `-I` flag of `protoc`). This means that a file located at `${SYS_PATH_ENTRY}/foo/bar/baz.proto` will result in the instantation of a module with the fully qualified name `foo.bar.baz_pb2`.  Users are expected to have a directory structure mirroring their desired import structure.
+
+Users have reported that getting Protobuf paths to match Python import paths is quite tricky today using `protoc`. In the case of a failure to import, the library will print troubleshooting information alongside the error.
+
+### Arbitrary Protos
+
+It should be stated that, in practice, it is possible to use these new functions to load totally arbitrary protos. Suppose you wrote a server that took `.proto` files as inputs from clients, instatiated modules from them, and returned some data about the file. For example, the number of message types contained within the supplied file. This could become problematic as new syntax features are added to the Protobuf specification. In the worst case, this would require a redeploy of the server with a sufficiently up-to-date version of `grpcio-tools`. But, regardless, we will claim no support for this use case. The intent of these functions is to enable import of *fixed* `.proto` files, known at build time.
+
+### Dependency Considerations
+
+gRPC makes a point not to incur a direct dependency on protocol buffers. It is not the intent of this feature to change that. Instead, the implementations of these new functions will live in the `grpcio-tools` package, which necessarily *already* has a hard dependency on `protobuf`. If the `grpcio` package finds that `grpc_tools` is importable, it will do so and use the implementations found there to back the `protos` and `services` functions. Otherwise, it will raise a `NotImplementedError`.
+
+In order to take advantage of newer language features, support will only be added for Python versions 3.6+. All other versions will result in a `NotImplementedError`.
+
+## Implementation
+
+This proposal has been implemented [here](https://github.com/grpc/grpc/pull/21458). This implementation uses the C++ protobuf runtime already built into the `grpcio-tools` C extension to parse the protocol buffers and generate textual Python code in memory. This code is then used to [instantiate the modules](https://www.python.org/dev/peps/pep-0302/) to be provided to the calling application.
+
```

Comment:
What version of protobuf runtime is supported in grpcio-tools and when is it upgraded?

Reply:
It's determined by the git submodule in the core repo. In other words, the same version as C++ at the same commit. Updating the submodule is now part of the gRPC Core release process.